cpp/src/arrow/util/hash-util.h

```diff
@@ -288,18 +365,49 @@ class HashUtil {
   }
 };
 
-// With sse4.2
+// Hash specialized for Arm Crc32
+template <>
+inline int HashUtil::Hash<USE_ARMCRC>(const void* data, int32_t bytes, uint32_t seed) {
+  // Need run time check
+  if (crc32c_runtime_check())
+    return static_cast<int>(HashUtil::CrcHashARMCE(data, bytes, seed));
+  else
+  // Fall back to Murmur
+    return static_cast<int>(HashUtil::MurmurHash2_64(data, bytes, seed));
+}
+
+
+// Hash specialized for sse4.2
 template <>
-inline int HashUtil::Hash<true>(const void* data, int32_t bytes, uint32_t seed) {
-  return static_cast<int>(HashUtil::CrcHash(data, bytes, seed));
+inline int HashUtil::Hash<USE_SSE42>(const void* data, int32_t bytes, uint32_t seed) {
+  return static_cast<int>(HashUtil::CrcHashSSE(data, bytes, seed));
 }
 
-// Non-sse4 hash
+// Hash specialized for default murmur hash
 template <>
-inline int HashUtil::Hash<false>(const void* data, int32_t bytes, uint32_t seed) {
+inline int HashUtil::Hash<USE_DEFAULT>(const void* data, int32_t bytes, uint32_t seed) {
   return static_cast<int>(HashUtil::MurmurHash2_64(data, bytes, seed));
 }
 
+// DoubleHash specialized for Arm Crc32
+template <>
+inline int HashUtil::DoubleHash<USE_ARMCRC>(const void* data, int32_t bytes, uint32_t seed) {
+  // Need run time check
+  if (crc32c_runtime_check())
```

Comment:
Just for the record, is the runtime check cheap? The hash function will often be called for smallish strings, e.g. 20 to 40 bytes.

Reply:
What does it mean for "cheap"? 
Sorry, I'm not familiar with the hash function calling scenario? 
If the hash function is called very frequently,  the run time check should be called just once to avoid performance degradation. I will try to move the `run time check` to a `init ` routine if possible.

Reply:
You may want to find out by running the `hashing-benchmark` with Arrow compiled in release mode. See if this PR actually gives a speedup :-)
(also, it would be nice to post the results so we non-ARM users can get an idea of the performance we're talking about)